clippy_lints/src/unwrap_in_result.rs

```diff
@@ -16,29 +16,36 @@ declare_clippy_lint! {
     ///
     /// **Example:**
     /// ```rust
-    /// // Bad
-    /// fn bad_divisible_by_3(i_str:String)->Result<bool,String> { // checks whether a string represents a number divisible by 3
-    ///     let i = i_str.parse::<i32>().unwrap();
-    ///     if i % 3 == 0 {
-    ///         Ok(true)
-    ///      } else {
-    ///         Err("Number is not divisible by 3")?
-    ///      }
-    /// }
-    /// Good
-    /// fn good_divisible_by_3(i_str:String)->Result<bool,String> {
-    ///     let i_result = i_str.parse::<i32>();
-    ///     match i_result
-    ///     {
-    ///         Ok(i) if i % 3 => Ok(true),
-    ///         Ok(_) => Err("Number is not divisible by 3")?
-    ///         Err(_) => Err("Not a number")?,
-    ///     },
+    /// struct A;
+    /// impl A {
+    ///     // Bad
+    ///     fn bad_divisible_by_3(i_str:String)->Result<bool,String> { // checks whether a string represents a number divisible by 3
+    ///         let i = i_str.parse::<i32>().unwrap();
+    ///         if i % 3 == 0 {
+    ///             Ok(true)
+    ///         } else {
+    ///             Err("Number is not divisible by 3")?
+    ///         }
+    ///     }
+    ///     // Good
+    ///     fn good_divisible_by_3(i_str:String)->Result<bool,String> {
+    ///         let i_result = i_str.parse::<i32>();
+    ///         match i_result
+    ///         {
+    ///             Err(_e) => Err("Not a number".to_string()),
+    ///             Ok(i) => {
+    ///                 if i % 3 == 0 {
+    ///                     return Ok(true);
+    ///                 }
+    ///                 Err("Number is not divisible by 3".to_string())
+    ///             },
+    ///         }
+    ///     }
     /// }
     /// ```
```

Comment:
it won't let me do a suggestion for this large of a code snippet but I'd like to replace the entire example with this revised version. The changes I've made are:

- switch `Ok(bool)` to a `Ok(())` because we never return an `Ok(false)`, so the boolean of "is divisible" is implied purely by the variant of the result via `divisible_by_three("6").is_ok()`
- used `map_err` instead of a custom match
- preserve the source error when creating the new error
- use early return for error case, rather than for Ok case, since an explicit return is already necessary to get the conversion to a String via `?`
- split the single example into two examples
- rename good / bad to before / after to avoid value judgements
- use the same function name for both examples
- applied rustfmt
- switch to expect to produce more consistent error messages across the two implementations
- changed higher level error message to indicate why the `divisible_by_3` function failed rather than why the `parse` function failed

```rust
/// Before:
/// ```rust
/// fn divisible_by_3(i_str: String) -> Result<(), String> {
///     let i = i_str
///         .parse::<i32>()
///         .expect("cannot divide the input by three");
///
///     if i % 3 != 0 {
///         Err("Number is not divisible by 3")?
///     }
///
///     Ok(())
/// }
/// ```
///
/// After:
/// ```rust
/// fn divisible_by_3(i_str: String) -> Result<(), String> {
///     let i = i_str
///         .parse::<i32>()
///         .map_err(|e| format!("cannot divide the input by three: {}", e))?;
///
///     if i % 3 != 0 {
///         Err("Number is not divisible by 3")?
///     }
///
///     Ok(())
/// }
/// ```
```